core/src/main/java/io/temporal/samples/earlyreturn/EarlyReturn.java

```diff
@@ -0,0 +1,256 @@
+        Workflow.newActivityStub(
+            TransactionActivities.class,
+            ActivityOptions.newBuilder().setStartToCloseTimeout(Duration.ofSeconds(30)).build());
+
+    private boolean initDone = false;
+    private Transaction tx;
+    private Exception initError = null;
+
+    @Override
+    public String processTransaction(Transaction txInput) {
+      this.tx = txInput;
+      // Phase 1: Initialize the transaction
+      try {
+        this.tx = activities.initTransaction(this.tx);
+      } catch (Exception e) {
+        initError = e;
+      } finally {
+        initDone = true;
+      }
+
+      // Phase 2: Complete or cancel the transaction
+      if (initError != null) {
+        activities.cancelTransaction(this.tx);
+        return "Transaction cancelled";
+      } else {
+        activities.completeTransaction(this.tx);
+        return "Transaction completed successfully: " + this.tx.id;
+      }
+    }
+
+    @Override
+    public String returnInitResult() {
+      Workflow.await(() -> initDone);
+      if (initError != null) {
+        log.info("Initialization failed.");
+        throw Workflow.wrap(initError);
+      }
+      return tx.getId();
+    }
+  }
+
+  @ActivityInterface
+  public interface TransactionActivities {
+    @ActivityMethod
+    Transaction initTransaction(Transaction tx);
+
+    @ActivityMethod
+    void cancelTransaction(Transaction tx);
+
+    @ActivityMethod
+    void completeTransaction(Transaction tx);
+  }
+
+  public static class TransactionActivitiesImpl implements TransactionActivities {
+    @Override
+    public Transaction initTransaction(Transaction tx) {
+      System.out.println("Initializing transaction");
+      sleep(500);
+      if (tx.getAmount() <= 0) {
+        System.out.println("Invalid amount: " + tx.getAmount());
+        throw ApplicationFailure.newNonRetryableFailure(
+            "Non-retryable Activity Failure: Invalid Amount", "InvalidAmount");
+      }
+      // mint a transaction ID
+      String transactionId =
+          "TXID" + String.format("%010d", (long) (Math.random() * 1_000_000_0000L));
+      tx.setId(transactionId);
+      sleep(500);
+      return tx;
+    }
+
+    @Override
+    public void cancelTransaction(Transaction tx) {
+      System.out.println("Cancelling transaction");
+      sleep(2000);
+      System.out.println("Transaction cancelled");
+    }
+
+    @Override
+    public void completeTransaction(Transaction tx) {
+      System.out.println(
+          "Sending $"
+              + tx.getAmount()
+              + " from "
+              + tx.getSourceAccount()
+              + " to "
+              + tx.getTargetAccount());
+      sleep(2000);
+      System.out.println("Transaction completed successfully");
+    }
+
+    private void sleep(long millis) {
+      try {
+        Thread.sleep(millis);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  public static class Transaction {
```

Comment:
For jackson serialization we should use annotations so we don't need to expose a bunch of extra method for serialization.

We have one example here https://github.com/temporalio/samples-java/blob/main/core/src/main/java/io/temporal/samples/nexus/service/NexusService.java#L62C5-L62C17

Oracle blog recommend the same approach https://blogs.oracle.com/javamagazine/post/java-json-serialization-jackson

Reply:
Got it, will implement. Thanks